File: javascript/iroha_js/src/connectSession.js
import { Buffer } from "node:buffer";
import { generateKeyPairSync, randomBytes } from "node:crypto";
import { blake2b } from "@noble/hashes/blake2b";
import { networkIdBytes } from "./networkId.js";

const SID_PREFIX = Buffer.from("iroha-connect|sid|");
const SID_LENGTH = 32;
const NONCE_LENGTH = 16;
const X25519_KEY_LENGTH = 32;
const CONNECT_URI_VERSION = "1";
const CONNECT_URI_SCHEME = "iroha://connect";
const CONNECT_RESPONSE_FIELDS = new Set([
  "sid",
  "network_id",
  "app_pk",
  "nonce",
  "wallet_uri",
  "app_uri",
  "token_app",
  "token_wallet",
  "token_management",
  "token_relay",
]);
const CONNECT_URI_FIELDS = new Set([
  "sid",
  "network_id",
  "app_pk",
  "nonce",
  "node",
  "v",
  "role",
  "token",
  "relay",
]);
const CONNECT_TOKEN_PATTERN = /^[A-Za-z0-9_-]{43}$/;

/**
 * Generate a Connect session identifier deterministically.
 * @param {{ networkId: import("./networkId.js").NetworkId; appPublicKey: BinaryLike; nonce?: BinaryLike | null }} options
 * @returns {{ sidBytes: Buffer; sidBase64Url: string; nonce: Buffer }}
 */
export function generateConnectSid(options = {}) {
  if (!options || typeof options !== "object") {
    throw new TypeError("options must be an object");
  }
  const networkId = options.networkId;
  const exactNetworkId = networkIdBytes(networkId, "networkId");
  const publicKey = normalizeBinary(options.appPublicKey, "appPublicKey", X25519_KEY_LENGTH);
  const nonce =
    options.nonce === undefined || options.nonce === null
      ? randomBytes(NONCE_LENGTH)
      : normalizeBinary(options.nonce, "nonce", NONCE_LENGTH);
  const digest = blake2b(
    Buffer.concat([SID_PREFIX, Buffer.from(exactNetworkId), publicKey, nonce]),
    { dkLen: SID_LENGTH },
  );
  const sidBytes = Buffer.from(digest);
  return {
    sidBytes,
    sidBase64Url: toBase64Url(sidBytes),
    nonce,
  };
}

/**
 * Create a Connect session preview by minting an X25519 keypair, nonce, and session URIs.
 * @param {{ networkId: import("./networkId.js").NetworkId; node?: string | null; nonce?: BinaryLike | null; appKeyPair?: { publicKey: BinaryLike; privateKey: BinaryLike } }} options
 * @returns {{
 *   networkId: import("./networkId.js").NetworkId;
 *   node: string | null;
 *   sidBytes: Buffer;
 *   sidBase64Url: string;
 *   nonce: Buffer;
 *   appKeyPair: { publicKey: Buffer; privateKey: Buffer };
 *   walletUri: string;
 *   appUri: string;
 * }}
 */
export function createConnectSessionPreview(options = {}) {
  if (!options || typeof options !== "object") {
    throw new TypeError("options must be an object");
  }
  const networkId = options.networkId;
  networkIdBytes(networkId, "networkId");
  const node =
    options.node === undefined || options.node === null
      ? null
      : requireNonEmptyString(options.node, "node");
  const appKeyPair = normalizeKeyPair(options.appKeyPair);
  const nonce =
    options.nonce === undefined || options.nonce === null
      ? randomBytes(NONCE_LENGTH)
      : normalizeBinary(options.nonce, "nonce", NONCE_LENGTH);
  const sidResult = generateConnectSid({
    networkId,
    appPublicKey: appKeyPair.publicKey,
    nonce,
  });
  return {
    networkId,
    node,
    sidBytes: sidResult.sidBytes,
    sidBase64Url: sidResult.sidBase64Url,
    nonce,
    appKeyPair,
    walletUri: buildConnectUri(
      sidResult.sidBase64Url,
      networkId,
      appKeyPair.publicKey,
      nonce,
      node,
      "wallet",
    ),
    appUri: buildConnectUri(
      sidResult.sidBase64Url,
      networkId,
      appKeyPair.publicKey,
      nonce,
      node,
      "app",
    ),
  };
}

/**
 * Fail closed when Torii substitutes any committed session field or deep-link
 * authorization value.
 * @param {Record<string, unknown>} session Parsed response (or a normalized
 * response carrying the original object as `raw`).
 * @param {{sid: string; networkId: string; appPk: string; nonce: string; node?: string | null}} expected
 */
export function validateConnectSessionResponseIdentity(session, expected) {
  if (!session || typeof session !== "object" || !expected || typeof expected !== "object") {
    throw new TypeError("Connect session response and expected identity must be objects");
  }
  const raw = session.raw && typeof session.raw === "object" ? session.raw : session;
  for (const field of Object.keys(raw)) {
    if (!CONNECT_RESPONSE_FIELDS.has(field)) {
      throw new Error(`Connect session response contains unsupported field ${field}`);
    }
  }
  const identity = {
    sid: requireExactString(raw.sid, "session.sid"),
    networkId: requireExactString(raw.network_id, "session.network_id"),
    appPk: requireExactString(raw.app_pk, "session.app_pk"),
    nonce: requireExactString(raw.nonce, "session.nonce"),
  };
  if (
    identity.sid !== expected.sid
    || identity.networkId !== expected.networkId
    || identity.appPk !== expected.appPk
    || identity.nonce !== expected.nonce
  ) {
    throw new Error("Torii substituted the canonical Connect session identity");
  }
  const tokens = {
    app: requireConnectToken(raw.token_app, "session.token_app"),
    wallet: requireConnectToken(raw.token_wallet, "session.token_wallet"),
    management: requireConnectToken(raw.token_management, "session.token_management"),
    relay: requireConnectToken(raw.token_relay, "session.token_relay"),
  };
  validateConnectLaunchUri(raw.wallet_uri, "wallet", identity, expected.node, tokens.wallet, tokens.relay);
  validateConnectLaunchUri(raw.app_uri, "app", identity, expected.node, tokens.app, tokens.relay);
}

function validateConnectLaunchUri(value, role, identity, node, token, relay) {
  const literal = requireExactString(value, `session.${role}_uri`);
  let parsed;
  try {
    parsed = new URL(literal);
  } catch {
    throw new Error(`session.${role}_uri must be an absolute Connect URI`);
  }
  if (
    parsed.protocol !== "iroha:"
    || parsed.host !== "connect"
    || parsed.pathname !== ""
    || parsed.hash !== ""
  ) {
    throw new Error(`session.${role}_uri must use iroha://connect without a path or fragment`);
  }
  const query = new Map();
  for (const [key, entry] of parsed.searchParams) {
    if (!CONNECT_URI_FIELDS.has(key)) {
      throw new Error(`session.${role}_uri contains unsupported parameter ${key}`);
    }
    if (query.has(key)) {
      throw new Error(`session.${role}_uri contains duplicate parameter ${key}`);
    }
    query.set(key, entry);
  }
  for (const field of CONNECT_URI_FIELDS) {
    if (!query.has(field)) {
      throw new Error(`session.${role}_uri is missing required parameter ${field}`);
    }
  }
  const expectedValues = {
    sid: identity.sid,
    network_id: identity.networkId,
    app_pk: identity.appPk,
    nonce: identity.nonce,
    node: node ?? "",
    v: CONNECT_URI_VERSION,
    role,
    token,
    relay,
  };
  for (const [field, expectedValue] of Object.entries(expectedValues)) {
    if (query.get(field) !== expectedValue) {
      throw new Error(`session.${role}_uri substituted Connect parameter ${field}`);
    }
  }
}

function requireExactString(value, name) {
  if (typeof value !== "string" || value.length === 0 || value.trim() !== value) {
    throw new TypeError(`${name} must be an exact non-empty string`);
  }
  return value;
}

function requireConnectToken(value, name) {
  const token = requireExactString(value, name);
  if (!CONNECT_TOKEN_PATTERN.test(token)) {
    throw new TypeError(`${name} must be a canonical 32-byte unpadded base64url token`);
  }
  return token;
}

function normalizeKeyPair(pair) {
  if (!pair) {
    const generated = generateX25519KeyPair();
    return {
      publicKey: generated.publicKey,
      privateKey: generated.privateKey,
    };
  }
  if (typeof pair !== "object") {
    throw new TypeError("appKeyPair must be an object");
  }
  return {
    publicKey: normalizeBinary(pair.publicKey, "appKeyPair.publicKey", X25519_KEY_LENGTH),
    privateKey: normalizeBinary(pair.privateKey, "appKeyPair.privateKey", X25519_KEY_LENGTH),
  };
}

function generateX25519KeyPair() {
  const { publicKey, privateKey } = generateKeyPairSync("x25519");
  const jwkPublic = publicKey.export({ format: "jwk" });
  const jwkPrivate = privateKey.export({ format: "jwk" });
  if (!jwkPublic?.x || !jwkPrivate?.d) {
    throw new Error("Failed to export x25519 key material");
  }
  return {
    publicKey: Buffer.from(jwkPublic.x, "base64url"),
    privateKey: Buffer.from(jwkPrivate.d, "base64url"),
  };
}

function normalizeConnectRole(role, name = "role") {
  if (role === "app" || role === "wallet") {
    return role;
  }
  throw new TypeError(`${name} must be 'app' or 'wallet'`);
}

function buildConnectUri(sidBase64Url, networkId, appPublicKey, nonce, node, role) {
  const params = new URLSearchParams();
  params.set("sid", sidBase64Url);
  params.set("network_id", networkId.toString());
  params.set("app_pk", toBase64Url(appPublicKey));
  params.set("nonce", toBase64Url(nonce));
  if (node) {
    params.set("node", node);
  }
  params.set("v", CONNECT_URI_VERSION);
  params.set("role", normalizeConnectRole(role));
  return `${CONNECT_URI_SCHEME}?${params.toString()}`;
}

function requireNonEmptyString(value, name) {
  if (typeof value !== "string") {
    throw new TypeError(`${name} must be a string`);
  }
  const trimmed = value.trim();
  if (!trimmed) {
    throw new Error(`${name} must not be empty`);
  }
  return trimmed;
}

function normalizeBinary(value, name, expectedLength) {
  const buffer = toBuffer(value, name);
  if (expectedLength !== undefined && buffer.length !== expectedLength) {
    throw new RangeError(
      `${name} must be ${expectedLength} bytes (received ${buffer.length} bytes)`,
    );
  }
  return buffer;
}

function normalizeByteArray(value, name) {
  const bytes = Array.from(value);
  const normalized = bytes.map((entry, index) => {
    if (!Number.isInteger(entry) || entry < 0 || entry > 0xff) {
      throw new TypeError(`${name}[${index}] must be a byte`);
    }
    return entry;
  });
  return Buffer.from(normalized);
}

function toBuffer(value, name) {
  if (Buffer.isBuffer(value)) {
    return Buffer.from(value);
  }
  if (ArrayBuffer.isView(value)) {
    return Buffer.from(value.buffer, value.byteOffset, value.byteLength);
  }
  if (value instanceof ArrayBuffer) {
    return Buffer.from(value);
  }
  if (Array.isArray(value)) {
    return normalizeByteArray(value, name);
  }
  if (typeof value === "string") {
    return decodeStringBinary(value, name);
  }
  if (value && typeof value.length === "number") {
    return normalizeByteArray(value, name);
  }
  throw new TypeError(`${name} must be binary data`);
}

function decodeStringBinary(input, name) {
  const trimmed = input.trim();
  if (!trimmed) {
    throw new TypeError(`${name} must not be empty`);
  }
  const hexPrefixed = trimmed.startsWith("0x") || trimmed.startsWith("0X");
  const hexBody = hexPrefixed ? trimmed.slice(2) : trimmed;
  if (/^[0-9a-fA-F]+$/.test(hexBody) && hexBody.length % 2 === 0) {
    return Buffer.from(hexBody, "hex");
  }
  try {
    return decodeBase64UrlStrict(trimmed, name);
  } catch {
    throw new TypeError(`${name} must be hex or base64 data`);
  }
}

function decodeBase64UrlStrict(value, name) {
  const normalized = value.replace(/-/g, "+").replace(/_/g, "/");
  let padded = normalized;
  const paddingIndex = normalized.indexOf("=");
  if (paddingIndex !== -1) {
    const head = normalized.slice(0, paddingIndex);
    const padding = normalized.slice(paddingIndex);
    if (!/^[0-9A-Za-z+/]*$/.test(head) || !/^={1,2}$/.test(padding)) {
      throw new TypeError(`${name} must be hex or base64 data`);
    }
    if (normalized.length % 4 !== 0) {
      throw new TypeError(`${name} must be hex or base64 data`);
    }
  } else {
    if (!/^[0-9A-Za-z+/]+$/.test(normalized) || normalized.length % 4 === 1) {
      throw new TypeError(`${name} must be hex or base64 data`);
    }
    const padLength = (4 - (normalized.length % 4)) % 4;
    padded = normalized + "=".repeat(padLength);
  }
  const decoded = Buffer.from(padded, "base64");
  if (decoded.toString("base64") !== padded) {
    throw new TypeError(`${name} must be hex or base64 data`);
  }
  return decoded;
}

function toBase64Url(buffer) {
  return Buffer.from(buffer)
    .toString("base64")
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/g, "");
}

File: python/iroha_torii_client/connect_session.py
"""Exact Connect session request and response validation."""

from __future__ import annotations

import base64
import binascii
import hashlib
import re
from typing import Any, Callable, Dict, Mapping, Tuple
from urllib.parse import parse_qsl, urlsplit

from .client_status_models import ConnectSessionInfo

HashLiteralValidator = Callable[[Any, str], str]

_SESSION_REQUEST_FIELDS = frozenset({"sid", "network_id", "app_pk", "nonce"})
_SESSION_OPTIONAL_FIELDS = frozenset({"node"})
_URI_FIELDS = frozenset(
    {"sid", "network_id", "app_pk", "nonce", "node", "v", "role", "token", "relay"}
)


def _require_exact_non_empty_string(value: Any, context: str) -> str:
    if not isinstance(value, str):
        raise TypeError(f"{context} must be a string")
    stripped = value.strip()
    if not stripped:
        raise ValueError(f"{context} must be a non-empty string")
    if stripped != value:
        raise ValueError(f"{context} must not contain surrounding whitespace")
    return value


def _network_identity(
    value: Any,
    context: str,
    hash_literal: HashLiteralValidator,
) -> Tuple[str, bytes]:
    literal = _require_exact_non_empty_string(value, context)
    try:
        canonical = hash_literal(literal, context)
    except RuntimeError as exc:
        raise ValueError(str(exc)) from exc
    return canonical, bytes.fromhex(canonical[5:69])


def _base64url(value: Any, length: int, context: str) -> Tuple[str, bytes]:
    encoded = _require_exact_non_empty_string(value, context)
    if re.fullmatch(r"[A-Za-z0-9_-]+", encoded) is None:
        raise ValueError(f"{context} must be canonical unpadded base64url")
    try:
        decoded = base64.urlsafe_b64decode(encoded + "=" * (-len(encoded) % 4))
    except (binascii.Error, ValueError) as exc:
        raise ValueError(f"{context} must be canonical unpadded base64url") from exc
    canonical = base64.urlsafe_b64encode(decoded).rstrip(b"=").decode("ascii")
    if len(decoded) != length or canonical != encoded:
        raise ValueError(
            f"{context} must be canonical unpadded base64url for exactly {length} bytes"
        )
    return encoded, decoded


def canonical_connect_sid(value: Any, context: str = "Connect session sid") -> str:
    """Return a canonical unpadded base64url 32-byte session identifier."""

    return _base64url(value, 32, context)[0]


def canonical_connect_token(value: Any, context: str) -> str:
    """Return a canonical unpadded base64url 32-byte Connect bearer token."""

    return _base64url(value, 32, context)[0]


def normalize_connect_session_request(
    payload: Mapping[str, Any],
    *,
    hash_literal: HashLiteralValidator,
) -> Dict[str, str]:
    """Validate and canonicalize an exact Connect session request."""

    if not isinstance(payload, Mapping):
        raise TypeError("Connect session request must be a mapping")
    request = dict(payload)
    retired = {"chain", "chain_id", "chainId", "genesis_hash", "genesisHash"}.intersection(
        request
    )
    if retired:
        raise ValueError("chain identity aliases are retired; provide exact network_id")
    missing = _SESSION_REQUEST_FIELDS.difference(request)
    if missing:
        raise ValueError(f"Connect session request missing required fields: {sorted(missing)}")
    unsupported = set(request).difference(_SESSION_REQUEST_FIELDS | _SESSION_OPTIONAL_FIELDS)
    if unsupported:
        rendered = sorted(str(field) for field in unsupported)
        raise ValueError(f"Connect session request has unsupported fields: {rendered}")

    network_id, network_bytes = _network_identity(
        request["network_id"],
        "Connect session request.network_id",
        hash_literal,
    )
    sid, sid_bytes = _base64url(request["sid"], 32, "Connect session request.sid")
    app_pk, app_pk_bytes = _base64url(request["app_pk"], 32, "Connect session request.app_pk")
    nonce, nonce_bytes = _base64url(request["nonce"], 16, "Connect session request.nonce")
    if not any(app_pk_bytes) or not any(nonce_bytes):
        raise ValueError("Connect session app_pk and nonce must not be all zero")
    expected_sid = hashlib.blake2b(
        b"iroha-connect|sid|" + network_bytes + app_pk_bytes + nonce_bytes,
        digest_size=32,
    ).digest()
    if sid_bytes != expected_sid:
        raise ValueError("Connect session sid does not match network_id, app_pk, and nonce")

    normalized = {
        "sid": sid,
        "network_id": network_id,
        "app_pk": app_pk,
        "nonce": nonce,
    }
    if "node" in request:
        normalized["node"] = _require_exact_non_empty_string(
            request["node"], "Connect session request.node"
        )
    return normalized


def _session_uri_query(
    uri: Any,
    *,
    context: str,
    session: ConnectSessionInfo,
    role: str,
    token: str,
) -> Dict[str, str]:
    value = _require_exact_non_empty_string(uri, context)
    parsed = urlsplit(value)
    if (
        parsed.scheme != "iroha"
        or parsed.netloc != "connect"
        or parsed.path not in {"", "/"}
        or parsed.fragment
    ):
        raise ValueError(f"{context} must be an iroha://connect URI")
    try:
        pairs = parse_qsl(parsed.query, keep_blank_values=True, strict_parsing=True)
    except ValueError as exc:
        raise ValueError(f"{context} contains a malformed query") from exc
    keys = [key for key, _ in pairs]
    duplicates = sorted({key for key in keys if keys.count(key) > 1})
    if duplicates:
        raise ValueError(f"{context} contains duplicate parameters: {duplicates}")
    query = dict(pairs)
    if set(query) != _URI_FIELDS:
        missing = sorted(_URI_FIELDS.difference(query))
        unsupported = sorted(set(query).difference(_URI_FIELDS))
        raise ValueError(
            f"{context} has an inexact parameter set; missing={missing}, unsupported={unsupported}"
        )
    expected = {
        "sid": session.sid,
        "network_id": session.network_id,
        "app_pk": session.app_pk,
        "nonce": session.nonce,
        "v": "1",
        "role": role,
        "token": token,
        "relay": session.token_relay,
    }
    if any(query[field] != expected_value for field, expected_value in expected.items()):
        raise ValueError(f"{context} substituted Connect session identity or authorization")
    return query


def ensure_connect_session_matches_request(
    session: ConnectSessionInfo,
    request: Mapping[str, str],
) -> None:
    """Reject a Torii response that substitutes request identity or node data."""

    if any(
        (
            session.sid != request["sid"],
            session.network_id != request["network_id"],
            session.app_pk != request["app_pk"],
            session.nonce != request["nonce"],
        )
    ):
        raise ValueError("Torii Connect session response substituted request identity")
    expected_node = request.get("node", "")
    wallet_query = _session_uri_query(
        session.wallet_uri,
        context="connect session.wallet_uri",
        session=session,
        role="wallet",
        token=session.token_wallet,
    )
    app_query = _session_uri_query(
        session.app_uri,
        context="connect session.app_uri",
        session=session,
        role="app",
        token=session.token_app,
    )
    if wallet_query["node"] != expected_node or app_query["node"] != expected_node:
        raise ValueError("Torii Connect session response substituted request node")


def parse_connect_session(
    payload: Mapping[str, Any],
    *,
    context: str,
    hash_literal: HashLiteralValidator,
) -> ConnectSessionInfo:
    """Parse and validate an exact Connect session response."""

    if not isinstance(payload, Mapping):
        raise RuntimeError(f"{context} response must be a JSON object")
    record = payload
    known = {
        "sid",
        "network_id",
        "app_pk",
        "nonce",
        "wallet_uri",
        "app_uri",
        "token_app",
        "token_wallet",
        "token_management",
        "token_relay",
    }
    if set(record) != known:
        missing = sorted(known.difference(record))
        unsupported = sorted(set(record).difference(known))
        raise ValueError(
            f"{context} response has an inexact field set; "
            f"missing={missing}, unsupported={unsupported}"
        )
    identity = normalize_connect_session_request(
        {
            "sid": record.get("sid"),
            "network_id": record.get("network_id"),
            "app_pk": record.get("app_pk"),
            "nonce": record.get("nonce"),
        },
        hash_literal=hash_literal,
    )
    wallet_uri = _require_exact_non_empty_string(
        record.get("wallet_uri"), f"{context}.wallet_uri"
    )
    app_uri = _require_exact_non_empty_string(record.get("app_uri"), f"{context}.app_uri")
    token_app = _base64url(record.get("token_app"), 32, f"{context}.token_app")[0]
    token_wallet = _base64url(record.get("token_wallet"), 32, f"{context}.token_wallet")[0]
    token_management = _base64url(
        record.get("token_management"), 32, f"{context}.token_management"
    )[0]
    token_relay = _base64url(record.get("token_relay"), 32, f"{context}.token_relay")[0]
    session = ConnectSessionInfo(
        sid=identity["sid"],
        network_id=identity["network_id"],
        app_pk=identity["app_pk"],
        nonce=identity["nonce"],
        wallet_uri=wallet_uri,
        app_uri=app_uri,
        token_app=token_app,
        token_wallet=token_wallet,
        token_management=token_management,
        token_relay=token_relay,
    )
    wallet_query = _session_uri_query(
        wallet_uri,
        context=f"{context}.wallet_uri",
        session=session,
        role="wallet",
        token=token_wallet,
    )
    app_query = _session_uri_query(
        app_uri,
        context=f"{context}.app_uri",
        session=session,
        role="app",
        token=token_app,
    )
    if wallet_query["node"] != app_query["node"]:
        raise ValueError(f"{context} response URIs disagree on node")
    return session
